tracer/src/Datadog.Trace/AsyncLocalScopeManager.cs

```diff
@@ -56,5 +57,27 @@ public void Close(Scope scope)
             // scope.Parent is null for distributed traces, so use scope.Span.Context.Parent
             DistributedTracer.Instance.SetSpanContext(scope.Span.Context.Parent as SpanContext);
         }
+
+        private static AsyncLocal<Scope> CreateScope()
+        {
+            if (Profiler.Instance.ContextTracker.IsEnabled)
+            {
+                return new AsyncLocal<Scope>(OnScopeChanged);
+            }
+
+            return new AsyncLocal<Scope>();
+        }
+
+        private static void OnScopeChanged(AsyncLocalValueChangedArgs<Scope> obj)
+        {
+            if (obj.CurrentValue == null)
```

Comment:
Is there a chance for the `obj.CurrentValue` to change concurrently to the execution of the method? I don't remember what's the usual semantic of callbacks like this in .NET are.

Reply:
I do not think that `obj.CurrentValue` can change concurrently to the execution of the method.
This callback is called by the thread itself.
 I'll confirm that.

Reply:
From what I see:
The threadpool queues (locals and global) have `IThreadPoolWorkItem` objects. Each time a threadpool thread pop an item, it will call the `ExecuteWorkItem` method on that item. In different implementation of that interface/method (Task, QueueUserWorkItemCallback, AwaitTakContinuation..), there is a call to `ExecutionContext.Run` (if there is any context).

In `ExecutionContext.Run`, if deemed necessary, `SetExecutionContext` is called. In this method, the [callback is called](https://referencesource.microsoft.com/#mscorlib/system/threading/executioncontext.cs,736).

Reply:
By convention, the event handler arguments (`EventArgs`) should not change during execution of the callback. It would be very troubling if it did!

In this case in particular, [`AsyncLocalValueChangedArgs<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocalvaluechangedargs-1) is a `readonly struct` so it is immutable.